sdk/opa.go

```diff
@@ -0,0 +1,336 @@
+	}
+
+	return opa, nil
+}
+
+// Start asynchronously starts the policy engine's plugins that download
+// policies, report status, etc.
+func (opa *OPA) Start(ctx context.Context) error {
+	return opa.manager.Start(ctx)
+}
+
+// Stop asynchronously stops the policy engine's plugins.
+func (opa *OPA) Stop(ctx context.Context) {
+	opa.manager.Stop(ctx)
+}
+
+// Query sends a query to the OPA instance and returns the decision
+func (opa *OPA) Query(ctx context.Context, decisionPath string, input interface{}, opts ...func(*rego.Rego)) (decision *Decision, err error) {
+	decision, err = newDecision()
+	if err != nil {
+		return nil, err
+	}
+
+	decision.Metrics.Timer("sdk_query").Start()
+	defer decision.Metrics.Timer("sdk_query").Stop()
+
+	var bundles map[string]server.BundleInfo
+
+	err = storage.Txn(ctx, opa.manager.Store, storage.TransactionParams{}, func(txn storage.Transaction) error {
+		var err error
+
+		if err := opa.getPreparedQuery(decisionPath, txn, decision.Metrics, opts...); err != nil {
+			return err
+		}
+		pq := opa.preparedQueries[decisionPath]
+
+		rs, err := pq.Eval(
+			ctx,
+			rego.EvalInput(input),
+			rego.EvalTransaction(txn),
+			rego.EvalMetrics(decision.Metrics),
+			rego.EvalInterQueryBuiltinCache(opa.interQueryBuiltinCache),
+		)
+
+		if err != nil {
+			return err
+		} else if len(rs) == 0 {
+			return fmt.Errorf("undefined decision")
+		} else {
+			decision.Data = rs[0].Expressions[0].Value
+		}
+
+		bundles, err = opa.bundles(ctx, txn)
+		if err != nil {
+			return fmt.Errorf("unexpected error %w", err)
+		}
+
+		return nil
+	})
+
+	if logger := logs.Lookup(opa.manager); logger != nil {
+		record := &server.Info{
+			Bundles:    bundles,
+			DecisionID: decision.ID,
+			Timestamp:  time.Now(),
+			Query:      decisionPath,
+			Input:      &input,
+			Error:      err,
+			Metrics:    decision.Metrics,
+		}
+		if err == nil {
+			var x = decision.Data
+			record.Results = &x
+		}
+
+		if err := logger.Log(ctx, record); err != nil {
+			return nil, fmt.Errorf("failed to log decision: %w", err)
+		}
+	}
+
+	return decision, err
+}
+
+// QueryDefault sends a query to the default decision path of the OPA instance and returns the decision
+func (opa *OPA) QueryDefault(ctx context.Context, input interface{}, opts ...func(*rego.Rego)) (decision *Decision, err error) {
+	decisionPath := *opa.manager.Config.DefaultDecision
+
+	return opa.Query(ctx, "data"+strings.ReplaceAll(decisionPath, "/", "."), input, opts...)
+}
+
+// AwaitReady waits for plugins and bundles to be loaded or until timeout on context is reached. Returns ready state.
+func (opa *OPA) AwaitReady(ctx context.Context) (ready bool) {
+	for {
+		pluginStatuses := opa.manager.PluginStatus()
+
+		if server.PluginsReady(pluginStatuses) && server.BundlesReady(pluginStatuses) {
+			ready = true
+			break
+		}
+
+		select {
+		case <-ctx.Done():
+			break
+		default:
+			time.Sleep(100 * time.Millisecond)
+		}
```

Comment:
This is fine. It's not uncommon to use [a ticker](https://golang.org/pkg/time/#Timer) here, too, I think:
```go
ticker := time.NewTicker(100 * time.Millisecond)
defer ticker.Stop()
select {
case <-ctx.Done():
    break
case <-ticker.C:
}
```
I believe the only difference is that the ticker-based approach will run the plugin status check *every 100ms*; the code we have here will let 100ms elapse *between checks*.

Reply:
👍 also found this now, perhaps it'd be worth re-using:

https://github.com/open-policy-agent/opa/blame/ce2aeaf163de16544e4d6923397442fbf017e1a3/util/wait.go#L16-L34

Reply:
Hmm I don't know. The "select ctx.Done" mechanism is basic enough to be used as-is, I'd say. Also, bridging the passed-in ctx with WaitFunc isn't trivial when considering explicit cancellation, too.

Let's keep it as-is 😃 

The only minor thing I'd change (it's a nitpick really) would be to return the ctx.Err() instead of what we create with fmt.Errorf here: 
https://github.com/open-policy-agent/opa/pull/3186/files#diff-9402e37dbcd2155047270666944796167deea341d9c6288dda653ca82b30ce87R176